specs/NavigateWithWebResourceRequest.md

```diff
@@ -0,0 +1,143 @@
+
+# Background
+Consumers of the old WebBrowser control that relied on the [Navigate](https://docs.microsoft.com/en-us/previous-versions/aa752133(v=vs.85)) API that allowed them to specify HTTP POST data and extra headers as part of navigation, [requested](https://github.com/MicrosoftEdge/WebViewFeedback/issues/69) same ability in WebView2. 
+
+# Description
+We propose adding NavigateWithWebResourceRequest to CoreWebView2 as the API that allows a WebView2 to navigate with a specified [CoreWebView2WebResourceRequest](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/dotnet/0-9-628/microsoft-web-webview2-core-corewebview2). This allows the developer to specify extra headers and POST data as part of their request for navigation.
+We also propose adding a CreateWebResourceRequest API to [CoreWebView2Environment](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/dotnet/0-9-515/microsoft-web-webview2-core-corewebview2environment) so that developers can create the WebResourceRequest object they'd like to use with NavigateWithWebResourceRequest.
+
+# Examples
+
+```cpp
+    // Need to convert post data to UTF-8 as required by the application/x-www-form-urlencoded Content-Type 
+    std::wstring postData = std::wstring(L"input=Hello");
+    int sizeNeededForMultiByte = WideCharToMultiByte(
+        CP_UTF8, 0, postData.c_str(), -1, nullptr,
+        0,
+        nullptr, nullptr);
+
+    char* postDataString = new char[sizeNeededForMultiByte];
+    WideCharToMultiByte(
+        CP_UTF8, 0, postData.c_str(), -1, postDataString, sizeNeededForMultiByte, nullptr,
+        nullptr);
+
+    wil::com_ptr<ICoreWebView2WebResourceRequest> webResourceRequest;
+    wil::com_ptr<IStream> postDataStream = SHCreateMemStream(
+        reinterpret_cast<const BYTE*>(postDataString), postData.length());
+
+    // This acts as a HTML form submit to https://www.w3schools.com/action_page.php
+    CHECK_FAILURE(webviewEnvironment->CreateWebResourceRequest(
+        L"https://www.w3schools.com/action_page.php", L"POST", postDataStream.get(),
+        L"Content-Type: application/x-www-form-urlencoded", &webResourceRequest));
+    CHECK_FAILURE(webview->NavigateWithWebResourceRequest(webResourceRequest.get()));
+```
+
+```csharp
+    UTF8Encoding utfEncoding = new UTF8Encoding();
+    byte[] postData = utfEncoding.GetBytes("input=Hello");
+
+    using (MemoryStream postDataStream = new MemoryStream(postData.Length))
+    {
+      postDataStream.Write(postData, 0, postData.Length);
+      CoreWebView2WebResourceRequest webResourceRequest = 
+        environment.CreateWebResourceRequest(
+          "https://www.w3schools.com/action_page.php",
+          "POST",
+          postDataStream,
```

Comment:
I believe SHCreateMemStream seeks to beginning, our APIs don't.

Reply:
SHCreateMemStream creates a stream initially positioned at offset 0. But in the C# version, we move the position as a result of the write. If callers need to rewind, then we should show the rewind. Otherwise, they'll copy this code, the posted data won't show up, and they'll assume there's an OS bug.